packages/marshal/src/helpers/environment-options.js

```diff
@@ -0,0 +1,119 @@
+
+const { details: X, quote: q } = assert;
+
+/**
+ * JavaScript module semantics resists attempts to parameterize a module's
+ * initialization behavior. A module initializes in order according to
+ * the path by which it is first imported, and then the initialized module
+ * is reused by all the other times it is imported. Compartments give us
+ * the opportunity to bind the same import name to different imported
+ * modules, depending on the package/compartment doing the import. Compartments
+ * do address the difficulty of parameterizing a module 's initialization
+ * logic, but not in a pleasant manner.
+ *
+ * A pleasant parameterization would be for a static module to be function-like
+ * with explicit parameters, and for the parameterization to be like
+ * calling the static module with parameters in order to derive from it a
+ * module instance. Compartments instead lets us parameterize the meaning
+ * of a module instance derived from a static module according to the
+ * three namespaces provided by the JavaScript semantics, effecting the
+ * meaning of a module instance.
+ *    * The global variable namespaces.
+ *       * The global scope, aliased to properties of the global object.
+ *         This is necessarily compartment-wide, and therefore in our
+ *         recommened usage pattern, packages - wide.
+ *       * The global lexical scope. The SES-shim compartments supports
+ *         these both compartment-wide as well as per-module. But it is
+ *         not yet clear what we will propose in the Compartment proposal.
+ *    * The import namespace.
+ *    * The host hooks.
+ *
+ * This `environment-options.js` module is a stopgap modeled on the treatment
+ * of the `LOCKDOWN_OPTIONS` configuration parameter as handled by
+ * `@agoric/install-ses`. It looks for a setting of this parameter in two
+ * places, both rooted in the global scope.
+ *    * It first looks for a global `LOCKDOWN_OPTIONS` variable which may
+ *      have already been set for the purpose, either by a module designed to
+ *      run first, or by a Compartment endowment. Indeed, Compartment
+ *      endowments are the most natural way Compartments provide to express
+ *      configuration parameters.
+ *    * Then for `process.env.LOCKDOWN_OPTIONS` rooted in the global
+ *      binding for `process` expected in the start compartment of Node
+ *      systems. In that context, `process.env` is expected to reflect
+ *      the Unix environment variables at the point that Node was
+ *      launched.
+ *
+ * Instead of looking for a global variable name per parameter, this
+ * `environment-options.js` instead first looks for a `ENV_OPTIONS` namespace
+ * that is expected to be like the `process.env` namespace expected on Node.
+ * If it finds that `ENV_OPTIONS` namespace at all, then the search stops there,
+ * whether on not the configuration parameter is found there. Only if
+ * `ENV_OPTIONS` does not exist will it then examine `process.env`.
```

Comment:
It’s perhaps cleaner to just lean into the convention established by Node.js such that we would endow compartments with a `process` where `class Process { env: Record<string, string> }`. It’s not the worst name for a namespace nor the worst shape, and would ease forward migration for a lot of code. This would also position us to anticipate the need to extend the shape to `class Process { env: Record<string, string>, argv: Array<string>, arg: string }`.

Reply:
I get your point. But `class`? WAT?

Reply:
Done! Resolving. PTAL